File: notion/qa/smart-capture.js
#!/usr/bin/env node
/**
 * Smart QA capture that matches Feedbucket screenshot region.
 * - Gets Feedbucket image dimensions for viewport sizing
 * - Detects if screenshot is from top of page (header visible)
 * - Uses pixel comparison for scroll position matching
 */

import sharp from 'sharp';
import https from 'https';
import http from 'http';
import fs from 'fs';
import path from 'path';

/**
 * Download image from URL to local path
 */
async function downloadImage(url, outputPath) {
  return new Promise((resolve, reject) => {
    const protocol = url.startsWith('https') ? https : http;
    const file = fs.createWriteStream(outputPath);

    protocol.get(url, (response) => {
      if (response.statusCode === 301 || response.statusCode === 302) {
        // Handle redirect
        downloadImage(response.headers.location, outputPath)
          .then(resolve)
          .catch(reject);
        return;
      }

      response.pipe(file);
      file.on('finish', () => {
        file.close();
        resolve(outputPath);
      });
    }).on('error', (err) => {
      fs.unlink(outputPath, () => {});
      reject(err);
    });
  });
}

/**
 * Get image dimensions from URL or local path
 */
async function getImageDimensions(pathOrUrl) {
  let imagePath = pathOrUrl;

  if (pathOrUrl.startsWith('http')) {
    imagePath = '/tmp/feedbucket_dim_check.jpg';
    await downloadImage(pathOrUrl, imagePath);
  }

  const metadata = await sharp(imagePath).metadata();
  return {
    width: metadata.width,
    height: metadata.height
  };
}

/**
 * Detect if image shows top of page (header visible)
 * Looks for dark header bar in top portion of image
 */
async function detectHeaderVisible(pathOrUrl) {
  let imagePath = pathOrUrl;

  if (pathOrUrl.startsWith('http')) {
    imagePath = '/tmp/feedbucket_header_check.jpg';
    await downloadImage(pathOrUrl, imagePath);
  }

  // Extract top 80 pixels and analyze
  const topRegion = await sharp(imagePath)
    .extract({ left: 0, top: 0, width: 1000, height: 80 })
    .grayscale()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const { data, info } = topRegion;

  // Count dark pixels (typical for header bars)
  let darkPixels = 0;
  for (let i = 0; i < data.length; i++) {
    if (data[i] < 60) darkPixels++;
  }

  const darkRatio = darkPixels / data.length;

  // Header typically present if more than 5% of top region is dark
  return darkRatio > 0.05;
}

/**
 * Find scroll position by comparing template to full page screenshot
 * Uses a simplified pixel matching approach
 */
async function findScrollPosition(fullPagePath, templatePathOrUrl) {
  let templatePath = templatePathOrUrl;

  if (templatePathOrUrl.startsWith('http')) {
    templatePath = '/tmp/feedbucket_scroll_template.jpg';
    await downloadImage(templatePathOrUrl, templatePath);
  }

  // Get dimensions
  const fullMeta = await sharp(fullPagePath).metadata();
  const tmplMeta = await sharp(templatePath).metadata();

  // Resize template to match full page width
  const scaledTemplate = await sharp(templatePath)
    .resize(fullMeta.width, null)
    .grayscale()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const scaledHeight = scaledTemplate.info.height;

  // Get full page as grayscale
  const fullPage = await sharp(fullPagePath)
    .grayscale()
    .raw()
    .toBuffer({ resolveWithObject: true });

  // Simple vertical scan to find best match
  let bestY = 0;
  let bestScore = Infinity;
  const stepSize = 10; // Check every 10 pixels for speed

  for (let y = 0; y <= fullMeta.height - scaledHeight; y += stepSize) {
    let diff = 0;
    const sampleStep = 100; // Sample every 100th pixel for speed

    for (let i = 0; i < scaledTemplate.data.length; i += sampleStep) {
      const fullIdx = (y * fullMeta.width) + (i % fullMeta.width) +
                      Math.floor(i / scaledTemplate.info.width) * fullMeta.width;
      diff += Math.abs(scaledTemplate.data[i] - (fullPage.data[fullIdx] || 0));
    }

    if (diff < bestScore) {
      bestScore = diff;
      bestY = y;
    }
  }

  return {
    scroll_y: bestY,
    confidence: 1 - (bestScore / (scaledTemplate.data.length / 100 * 255)),
    template_height: scaledHeight
  };
}

/**
 * Analyze what's needed to capture the same region as Feedbucket
 */
async function analyzeCaptureRequirements(feedbucketUrl, fullPagePath = null) {
  // Get Feedbucket image dimensions
  const dims = await getImageDimensions(feedbucketUrl);

  // Check if header is visible (top of page)
  const headerVisible = await detectHeaderVisible(feedbucketUrl);

  const result = {
    viewport_width: dims.width,
    viewport_height: dims.height,
    capture_type: headerVisible ? 'viewport' : 'scroll_match',
    header_visible: headerVisible
  };

  // If header not visible and we have a full page screenshot, find scroll position
  if (!headerVisible && fullPagePath) {
    const scrollInfo = await findScrollPosition(fullPagePath, feedbucketUrl);
    result.scroll_y = scrollInfo.scroll_y;
    result.match_confidence = scrollInfo.confidence;
  } else if (!headerVisible) {
    result.needs_full_page = true;
  }

  return result;
}

/**
 * Build recapture queue with viewport info for all tickets
 */
async function buildRecaptureQueue(inputQueuePath, outputQueuePath) {
  const queue = JSON.parse(fs.readFileSync(inputQueuePath, 'utf8'));

  const recaptureQueue = [];

  for (const item of queue) {
    const fbUrl = item.feedbucket_url;
    if (fbUrl) {
      try {
        const dims = await getImageDimensions(fbUrl);
        item.viewport_width = dims.width;
        item.viewport_height = dims.height;
        item.capture_type = 'viewport';
      } catch (err) {
        console.error(`Error getting dimensions for ${item.ticket_id}: ${err.message}`);
        item.viewport_width = 1440;
        item.viewport_height = 900;
        item.capture_type = 'viewport';
      }
    }
    recaptureQueue.push(item);
  }

  fs.writeFileSync(outputQueuePath, JSON.stringify(recaptureQueue, null, 2));

  // Group by page_url + viewport
  const groups = {};
  for (const item of recaptureQueue) {
    const key = `${item.page_url.split('?')[0]}|${item.viewport_width}x${item.viewport_height}`;
    if (!groups[key]) groups[key] = [];
    groups[key].push(item);
  }

  return { queue: recaptureQueue, groups };
}

// CLI interface
async function main() {
  const args = process.argv.slice(2);

  if (args.includes('--help') || args.length === 0) {
    console.log(`
Smart QA Capture Tool

Usage:
  node smart-capture.js --analyze --feedbucket-url <url> [--full-page <path>]
  node smart-capture.js --build-queue --input <queue.json> --output <recapture.json>
  node smart-capture.js --dimensions --feedbucket-url <url>

Options:
  --analyze           Analyze Feedbucket image for capture requirements
  --build-queue       Build recapture queue with viewport info
  --dimensions        Just get image dimensions
  --feedbucket-url    URL to Feedbucket image
  --full-page         Path to full page screenshot (for scroll detection)
  --input             Input queue JSON file
  --output            Output queue JSON file
`);
    process.exit(0);
  }

  try {
    if (args.includes('--dimensions')) {
      const urlIdx = args.indexOf('--feedbucket-url');
      const url = args[urlIdx + 1];
      const dims = await getImageDimensions(url);
      console.log(JSON.stringify(dims));
    }
    else if (args.includes('--analyze')) {
      const urlIdx = args.indexOf('--feedbucket-url');
      const url = args[urlIdx + 1];
      const fullPageIdx = args.indexOf('--full-page');
      const fullPage = fullPageIdx !== -1 ? args[fullPageIdx + 1] : null;

      const result = await analyzeCaptureRequirements(url, fullPage);
      console.log(JSON.stringify(result));
    }
    else if (args.includes('--build-queue')) {
      const inputIdx = args.indexOf('--input');
      const outputIdx = args.indexOf('--output');
      const inputPath = args[inputIdx + 1];
      const outputPath = args[outputIdx + 1];

      const { queue, groups } = await buildRecaptureQueue(inputPath, outputPath);

      console.log(`Generated recapture queue with ${queue.length} tickets`);
      console.log(`\nGrouped into ${Object.keys(groups).length} unique page+viewport combinations:`);

      const sorted = Object.entries(groups).sort((a, b) => b[1].length - a[1].length);
      for (const [key, items] of sorted) {
        console.log(`  ${key}: ${items.length} tickets`);
      }
    }
  } catch (err) {
    console.error(JSON.stringify({ error: err.message }));
    process.exit(1);
  }
}

main();

export {
  getImageDimensions,
  detectHeaderVisible,
  findScrollPosition,
  analyzeCaptureRequirements,
  buildRecaptureQueue
};

File: notion/qa/smart-capture.py
#!/usr/bin/env python3
"""
Smart QA capture that matches Feedbucket screenshot region.
- Detects if screenshot is from top of page (header visible)
- Uses OpenCV template matching for scrolled screenshots
- Returns capture instructions for Chrome MCP
"""

import cv2
import numpy as np
import sys
import argparse
import urllib.request
import json
from PIL import Image

def download_image(url, output_path):
    """Download image from URL"""
    urllib.request.urlretrieve(url, output_path)
    return output_path

def get_image_dimensions(path_or_url):
    """Get image width and height"""
    if path_or_url.startswith('http'):
        temp_path = '/tmp/feedbucket_check.jpg'
        download_image(path_or_url, temp_path)
        path_or_url = temp_path

    img = Image.open(path_or_url)
    return {'width': img.width, 'height': img.height}

def detect_header_visible(image_path_or_url, threshold=50):
    """
    Detect if the image shows the top of a page (header visible).
    Looks for dark header bar or navigation elements in top portion.
    """
    if image_path_or_url.startswith('http'):
        temp_path = '/tmp/feedbucket_detect.jpg'
        download_image(image_path_or_url, temp_path)
        image_path_or_url = temp_path

    img = cv2.imread(image_path_or_url)
    if img is None:
        return True  # Default to viewport capture

    height, width = img.shape[:2]

    # Check top 100 pixels for header-like elements
    top_region = img[0:min(100, height), :]

    # Convert to grayscale and look for dark regions (typical header)
    gray = cv2.cvtColor(top_region, cv2.COLOR_BGR2GRAY)

    # Check for significant dark regions (header bars are often dark)
    dark_pixels = np.sum(gray < 50)
    total_pixels = gray.size
    dark_ratio = dark_pixels / total_pixels

    # If more than 10% of top region is dark, likely has header
    # Also check for navigation-like horizontal lines
    edges = cv2.Canny(gray, 50, 150)
    horizontal_lines = np.sum(edges) / total_pixels

    # Header typically present if dark ratio > 0.1 or significant edges
    return dark_ratio > 0.05 or horizontal_lines > 10

def find_scroll_position(full_page_path, template_path_or_url):
    """
    Find where the template appears in the full page screenshot.
    Returns the Y scroll position needed.
    """
    full_page = cv2.imread(full_page_path)

    if template_path_or_url.startswith('http'):
        temp_path = '/tmp/feedbucket_template.jpg'
        download_image(template_path_or_url, temp_path)
        template = cv2.imread(temp_path)
    else:
        template = cv2.imread(template_path_or_url)

    if full_page is None or template is None:
        return {'scroll_y': 0, 'confidence': 0}

    # Get dimensions
    tmpl_h, tmpl_w = template.shape[:2]
    full_h, full_w = full_page.shape[:2]

    # Scale template to match full page width if needed
    if tmpl_w != full_w:
        scale = full_w / tmpl_w
        template = cv2.resize(template, (full_w, int(tmpl_h * scale)))
        tmpl_h, tmpl_w = template.shape[:2]

    # Convert to grayscale
    full_gray = cv2.cvtColor(full_page, cv2.COLOR_BGR2GRAY)
    tmpl_gray = cv2.cvtColor(template, cv2.COLOR_BGR2GRAY)

    # Skip if template is taller than full page
    if tmpl_h >= full_h:
        return {'scroll_y': 0, 'confidence': 0}

    # Template matching
    result = cv2.matchTemplate(full_gray, tmpl_gray, cv2.TM_CCOEFF_NORMED)
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)

    return {
        'scroll_y': max_loc[1],
        'confidence': max_val,
        'match_x': max_loc[0],
        'template_height': tmpl_h
    }

def analyze_capture_requirements(feedbucket_url, full_page_path=None):
    """
    Analyze what's needed to capture the same region as Feedbucket.
    Returns capture instructions.
    """
    # Get Feedbucket image dimensions
    dims = get_image_dimensions(feedbucket_url)

    # Check if header is visible (top of page)
    header_visible = detect_header_visible(feedbucket_url)

    result = {
        'viewport_width': dims['width'],
        'viewport_height': dims['height'],
        'capture_type': 'viewport' if header_visible else 'scroll_match',
        'header_visible': bool(header_visible)
    }

    # If header not visible and we have a full page screenshot, find scroll position
    if not header_visible and full_page_path:
        scroll_info = find_scroll_position(full_page_path, feedbucket_url)
        result['scroll_y'] = scroll_info['scroll_y']
        result['match_confidence'] = scroll_info['confidence']
    elif not header_visible:
        result['needs_full_page'] = True

    return result

def main():
    parser = argparse.ArgumentParser(description='Analyze Feedbucket screenshot for smart capture')
    parser.add_argument('--feedbucket-url', required=True, help='URL to Feedbucket image')
    parser.add_argument('--full-page', help='Path to full page screenshot (for scroll detection)')
    parser.add_argument('--output-instructions', help='Output JSON file for capture instructions')

    args = parser.parse_args()

    result = analyze_capture_requirements(args.feedbucket_url, args.full_page)

    if args.output_instructions:
        with open(args.output_instructions, 'w') as f:
            json.dump(result, f, indent=2)

    print(json.dumps(result))

if __name__ == '__main__':
    main()
